File: src/quran_align_player/parsing/ParsedAyah.java
package quran_align_player.parsing;

import java.util.List;

public class ParsedAyah {
    private int number;
    private final List<Segment> segments;
    private final SurahEntry.Stats stats;
    private final String text;

    ParsedAyah(int number, List<Segment> segments, SurahEntry.Stats stats, String text) {
        this.number = number;
        this.segments = segments;
        this.stats = stats;
        this.text = text;
    }

    int getNumber() {
        return number;
    }

    public Segment getSegmentAt(int millis) {
        for (Segment s : segments) {
            if (s.endMillis >= millis)
                return s;
        }

        // This current timestamp is bigger than all elements, return the largest
        return segments.get(segments.size() - 1);
    }

    public int getDeletions() {
        return stats.deletions;
    }

    public int getTranspositions() {
        return stats.transpositions;
    }

    public int getInsertions() {
        return stats.insertions;
    }

    public String getText() {
        return text;
    }
}

File: align_parser/src/align_parsing/ParsedAyah.kt
package align_parsing

internal class ParsedAyah
    (
    val number: Int, val segments: List<Segment>, private val stats: SurahEntry.Stats,
    val text: String
) {

    init {
        if (stats.deletions == 0) {
            val textWords: List<String> = text.split(" ").filter { it.isNotEmpty() }
            for (segment in segments) {
                val range = segment.startWordIndex until segment.endWordIndex
                segment.setText((textWords.slice(range).joinToString(" ")))
            }
        }
    }

    val deletions: Int
        get() = stats.deletions

    val transpositions: Int
        get() = stats.transpositions

    val insertions: Int
        get() = stats.insertions

    fun getSegmentAt(millis: Int): Segment {
        for (s in segments) {
            if (s.endMillis >= millis)
                return s
        }

        // This current timestamp is bigger than all elements, return the largest
        return segments[segments.size - 1]
    }
}
